torch/quantization/observer.py

```diff
@@ -1072,6 +1073,91 @@ def forward(self, x):
 
+def _is_observer_script_module(mod, obs_type_name):
+    ''' Returns true if given mod is an instance of Observer script module.
+    '''
+    if isinstance(mod, torch.jit.RecursiveScriptModule):
+        # qualified name looks like '__torch__.torch.quantization.observer.___torch_mangle_2.MinMaxObserver'
+        suffix = mod._c.qualified_name.split('.', 1)[1]
+        name = re.sub(r'\.___torch_mangle_\d+', '', suffix)
+        return obs_type_name in name
+    return False
+
+def _is_activation_post_process(module):
+    return (isinstance(module, torch.quantization.ObserverBase) or
+            isinstance(module, torch.quantization.FakeQuantize) or
+            _is_observer_script_module(module, 'torch.quantization.observer'))
+
+def _is_per_channel_obs_instance(module):
+    if isinstance(module, torch.jit.RecursiveScriptModule):
+        return _is_observer_script_module(module, "torch.quantization.observer.PerChannelMinMaxObserver") or\
+            _is_observer_script_module(module, "torch.quantization.observer.MovingAveragePerChannelMinMaxObserver")
+    else:
+        return isinstance(module, torch.quantization.PerChannelMinMaxObserver) or\
+            isinstance(module, torch.quantization.MovingAveragePerChannelMinMaxObserver)
+
+def _is_per_tensor_obs_instance(module):
+    if isinstance(module, torch.jit.RecursiveScriptModule):
+        return _is_observer_script_module(module, "torch.quantization.observer.MinMaxObserver")
+    else:
+        return isinstance(module, torch.quantization.MinMaxObserver)
+
+def _is_hist_obs_instance(module):
+    if isinstance(module, torch.jit.RecursiveScriptModule):
+        return _is_observer_script_module(module, "torch.quantization.observer.MinMaxObserver")
+    else:
+        return isinstance(module, torch.quantization.HistogramObserver)
+
+def _get_attr_names_for_obs(module, name):
+    r"""
+    Return a dict that contains the module attr to fully qualified name
+    in the model. The name should correspond to the attr name in the model
+    state_dict.
+    """
+    qual_names = {}
+    qual_names['eps'] = name + '.eps'
+    if _is_per_channel_obs_instance(module):
+        qual_names['min_vals'] = name + '.min_vals'
+        qual_names['max_vals'] = name + '.max_vals'
+    elif _is_per_tensor_obs_instance(module):
+        qual_names['min_val'] = name + '.min_val'
+        qual_names['max_val'] = name + '.max_val'
+    elif _is_hist_obs_instance(module):
+        qual_names['histogram'] = name + '.histogram'
+        qual_names['min_val'] = name + '.min_val'
+        qual_names['max_val'] = name + '.max_val'
+
+    return qual_names
+
+def get_observer_state_dict(mod):
+    r"""
+    Returns the state dict corresponding to the observer stats.
+    Traverse the model state_dict and extract out the stats.
+    """
+    od = OrderedDict()
+    if isinstance(mod, torch.jit.RecursiveScriptModule):
+        for k, v in mod.state_dict().items():
```

Comment:
would it be possible to return the state dict directly instead of copying it over?  it would reduce risk of bugs if there are any changes to how state dicts work and someone forgets to update this file

Reply:
You mean update the original state_dict?
> it would reduce risk of bugs if there are any changes to how state dicts work and someone forgets to update this file

Wouldn't the unit tests catch this?

> would the current code copy the version over as well?
Not right now, but I can add that.

Reply:
Versioning is quite important. lets add support for that!

Reply:
currently we seem to be inspecting the state dict and copying things over one by one.  If instead we treat the state_dict for a module as a black box, then we remove the surface area of potential future bugs which could happen when state dict format changes (because we do not inspect it).